### ISSUE TITLE
can't set Associations

### ISSUE THREAD

**Issue by dongshimou** [go-gorm/gorm#1824]
go version : `go version go1.9.2 darwin/amd64`

database : `mysql 5.7`

gorm : `master`

the error : `Unknown column 'authors.title' in 'where clause'`

code like this : 
```
import (
	"fmt"
	_ "github.com/go-sql-driver/mysql"
	"github.com/jinzhu/gorm"
	"log"
)

var (
	db *gorm.DB
)

type Article struct {
	gorm.Model
	Title string `gorm:"size:255;unique;"`
	Author Author `gorm:"FOREIGNKEY:AuthorID;"`
	AuthorID int
}
type Author struct {
	gorm.Model
	Email string `gorm:"size:255;unique;"`
}
func TestDB(){
	Username,Password,Host,Port,Database:="xxx","xxx","xxx","xxx","xxxx"
	args := fmt.Sprintf("%s:%s@tcp(%s:%s)/%s?charset=utf8&parseTime=True&loc=Local",
		Username, Password, Host, Port, Database)
	db, _ = gorm.Open("mysql", args)
	gorm.DefaultTableNameHandler = func(db *gorm.DB, defaultTableName string) string {
		return defaultTableName
	}
	db = db.Set("gorm:table_options", "ENGINE=InnoDB CHARSET=utf8 auto_increment=1")
	db.CreateTable(&Article{})
	db.CreateTable(&Author{})

	db.Create(&Article{
	Title:"test title",
		Author:Author{
		Email:"test@xx.com",
		},
	})

	art:=Article{}
	db=db.Where(&Article{Title:"test title"}).Find(&art)
	db=db.Related(&art.Author)
	// error happend
	if db.Error!=nil{
		log.Print(db.Error)
	}
}
```
when i trace stack.
`func (scope *Scope) related(value interface{}, foreignKeys ...string) *Scope `
the `fromField` and the `toField` is not exist at the same time...
and not found the `relationship`

## it is a bug?
@jinzhu 
thanks.

**Comment by lukaskj** [comment]
Same problem here.
Not working even with Preload (#1764)
Tried with _db.Preaload("TableName")_ and _.Set("gorm:auto_preload", true)_

1. In your example _AuthorID_ is with the wrong type (_int_). The correct type must be _uint_ if you are using the default _gorm.Model_ columns.
This didn't fix the issue (even with your example in my machine)


1. The workaround that I was using to fix this FK issue (and the reason that I stopped using gorm) is:
- Added the FKs by hand with
`db.Model(&Article{}).AddForeignKey("author_id", "authors(id)", "RESTRICT", "RESTRICT")`

The above solution don't work with SQLite databases. And I needed to use both (MySQL and SQLite).
To fix this issue with SQLite I declared a gorm:type in struct like:
```golang
type User struct {
	gorm.Model		
	Role   *Role `gorm:"foreignkey:RoleID"`
	RoleID uint  `gorm:"type:INT UNSIGNED REFERENCES roles(id) ON DELETE RESTRICT ON UPDATE RESTRICT"`
}
```

And just with both _workarounds_ that I could make FKs work in MySQL and SQLite at the same time.

I don't know if I am doing something wrong

Some docs that I followed to create relations: [GORM Docs: Belongs To](http://gorm.io/docs/belongs_to.html)

**Comment by lukaskj** [comment]
For your specific problem the fix is to use Preload (#1764)

1. In your example _AuthorID_ is with the wrong type (_int_). The correct type must be _uint_ if you are using the default _gorm.Model_ columns.

I found this issue running `SHOW ENGINE INNODB STATUS;` and looking into _LATEST FOREIGN KEY ERROR_ section

2. Added preload into _db_
```golang
db = db.Set("gorm:table_options", "ENGINE=InnoDB CHARSET=utf8 auto_increment=1").Set("gorm:auto_preload", true)
```
3. Removed `db = db.Related(&art.Author)`
With preload set you can ignore this _.Related_ and print the author to verify `fmt.Println(art.Author)`

But the foreign keys are not being created yet

----------
For the FK creation problem:

1. Added the FKs by hand with
`db.Model(&Article{}).AddForeignKey("author_id", "authors(id)", "RESTRICT", "RESTRICT")`

2. The above solution don't work with SQLite databases. And I needed to use both (MySQL and SQLite).
To fix this issue with SQLite I declared a gorm:type in struct like:
```golang
type User struct {
	gorm.Model		
	Role   *Role `gorm:"foreignkey:RoleID"`
	RoleID uint  `gorm:"type:INT UNSIGNED REFERENCES roles(id) ON DELETE RESTRICT ON UPDATE RESTRICT"`
}
```

And just with both _workarounds_ that I could make FKs work in MySQL and SQLite at the same time.

I don't know if I am doing something wrong

Some docs that I followed to create relations: [GORM Docs: Belongs To](http://gorm.io/docs/belongs_to.html)

**Comment by dongshimou** [comment]
@lukaskj thanks...

i use the `db=db.Related(&art.Author,"author_id")` to solve my problem.

but when i save the `gorm:"FOREIGNKEY:AuthorID"`, is it invalid?
or maybe must set `gorm:auto_preload`. 

i don't want manual setting all ForeignKey... 😂
